### PR TITLE
Secret providers

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

File: pkg/authenticator/authenticator.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	IBMCLOUD_PROFILEID        = "IBMCLOUD_PROFILEID"
+	IAM                       = "iam"
+	PODIDENTITY               = "pod-identity"
+)
+
+var defaultSecret string
+
+// Authenticator ...
+type Authenticator interface {
+	GetToken(freshTokenRequired bool) (string, uint64, error)
```
Comment (GunaKKIBM) [inline pkg/authenticator/authenticator.go:42]: Add reasonForCall.

File: pkg/authenticator/api-key.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	tokenlifetime, err = token.CheckTokenLifeTime(iamtoken)
+	if err != nil {
+		aa.logger.Error("Error fetching tokenlifetime", zap.Error(err))
+		return "", tokenlifetime, err
+	}
+
+	aa.logger.Info("Successfully fetched IAM token")
+	return tokenResponse.AccessToken, tokenlifetime, nil
```
Comment (GunaKKIBM) [inline pkg/authenticator/api-key.go:89]: Change to iamToken

File: pkg/authenticator/authenticator.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	receivedError := err
+	errMsg := strings.ToLower(err.Error())
+	switch authType {
+	case IAM:
+		if !strings.Contains(errMsg, utils.APIKeyNotFound) || !strings.Contains(errMsg, utils.UserNotFound) {
+			return err
+		}
+	case PODIDENTITY:
+		if !strings.Contains(errMsg, utils.ProfileNotFound) {
+			return err
```
Comment (GunaKKIBM) [inline pkg/authenticator/authenticator.go:151]: Formulate a proper error message.

File: pkg/authenticator/authenticator.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	for retryCount := 0; retryCount < utils.MaxRetries; retryCount++ {
+		credentialsmap, err := parseCredentials(logger)
+		if err != nil {
+			return err
+		}
+
+		retrievedAuthType, _ := credentialsmap[IBMCLOUD_AUTHTYPE]
+		if retrievedAuthType != authType {
```
Comment (GunaKKIBM) [inline pkg/authenticator/authenticator.go:162]: Handle a diff in auth type

**Review by arahamad — APPROVED** [review]
lgtm

**Pull request merged**